lib/paperclip/vips_lazy_thumbnail.rb

```diff
@@ -0,0 +1,143 @@
+# frozen_string_literal: true
+
+module Paperclip
+  class LazyThumbnail < Paperclip::Processor
+    GIF_FPS = 12
+
+    GIF_PALETTE_COLORS = 32
+
+    ALLOWED_FIELDS = %w(
+      icc-profile-data
+    ).freeze
+
+    class PixelGeometryParser
+      def self.parse(current_geometry, pixels)
+        width  = Math.sqrt(pixels * (current_geometry.width.to_f / current_geometry.height)).round.to_i
+        height = Math.sqrt(pixels * (current_geometry.height.to_f / current_geometry.width)).round.to_i
+
+        Paperclip::Geometry.new(width, height)
+      end
+    end
+
+    def initialize(file, options = {}, attachment = nil)
+      super
+
+      @crop = options[:geometry].to_s[-1, 1] == '#'
+      @current_geometry = options.fetch(:file_geometry_parser, Geometry).from_file(@file)
+      @target_geometry = options[:pixels] ? PixelGeometryParser.parse(@current_geometry, options[:pixels]) : options.fetch(:string_geometry_parser, Geometry).parse(options[:geometry].to_s)
+      @format = options[:format]
+      @current_format = File.extname(@file.path)
+      @basename = File.basename(@file.path, @current_format)
+
+      correct_current_format!
+    end
+
+    def make
+      return File.open(@file.path) unless needs_convert?
+
+      dst = TempfileFactory.new.generate([@basename, @format ? ".#{@format}" : @current_format].join)
+
+      if preserve_animation?
+        if @target_geometry.nil? || (@current_geometry.width <= @target_geometry.width && @current_geometry.height <= @target_geometry.height)
+          target_width = 'iw'
+          target_height = 'ih'
+        else
+          scale = [@target_geometry.width.to_f / @current_geometry.width, @target_geometry.height.to_f / @target_geometry.height].min
+          target_width = (@current_geometry.width * scale).round
+          target_height = (@current_geometry.height * scale).round
+        end
+
+        # The only situation where we use crop on GIFs is cropping them to a square
+        # aspect ratio, so this is the only special case we implement. If cropping
+        # ever becomes necessary for other situations, this will need to be expanded.
+        if @target_geometry&.square?
+          crop_width = [target_width, target_height].min
+          crop_height = [target_width, target_height].min
+        end
```

Comment:
In which case could the `@target_geometry` be square while the source isn't?

Reply:
Yes, absolutely. Sorry if this is written confusingly, crop width and height are set to the exact same value, what I would call `min_side`.

Reply:
Oh, I understand, this can happen when the target is an avatar. Rewrote it to make it clearer but the logic is sound.